### ISSUE TITLE
Specify output height/width directly in TransformerLayer (instead of scaling factor)

### ISSUE THREAD

**Issue by benanne** [Lasagne/Lasagne#399]
I think it would be more natural to specify the output width and height of a `TransformerLayer` directly, rather than specifying a downsampling factor. This is especially the case when the desired layer output has a different aspect ratio than the input. I realise that you can specify a tuple of two downsample factors for this case, but still, it seems a lot more straightforward to just specify the size in pixels directly.

A quick scan through the code shows that `downsample_factor` is only ever used to compute the output shape from the input shape anyway, so it doesn't seem like this makes any other part of the implementation more straightforward.

Was there any particular motivation to go with the scaling factor approach? If not, I think we should probably refactor the code to at least allow both approaches. Thoughts?

**Comment by f0k** [comment]
> Thoughts?

I don't know what the original motivation was either, but I find it convenient to be able to just specify a factor -- otherwise I'd have to know the input shape. The only thing I don't like is the long argument name `downscale_factor`, I'd have preferred `scale` (with the inverse meaning, so `0.5` would halve the size). We could still change that if we wanted, the `TransformerLayer` was added after the release.

Note that the output shape is computed symbolically, so the factor and the fixed output shape would result in different behaviour for variable input sizes (the factor would adapt to the size, the fixed shape would be, well, fixed).

> If not, I think we should probably refactor the code to at least allow both approaches.

Sure, allowing both approaches wouldn't be a degradation. It's just that you'd need to distinguish the two cases in `get_output_shape_for()` and `get_output_for()`, you can't just solve that in the constructor already.

**Comment by benanne** [comment]
> the factor and the fixed output shape would result in different behaviour for variable input sizes (the factor would adapt to the size, the fixed shape would be, well, fixed).

But a fixed shape is what you want, no? Except in the special case where the network you stack on top of it is fully convolutional and you have some kind of global pooling layer at the end.

So I don't really understand how being able to specify a scale factor exempts you from knowing the input shape -- you still need to know what shape the output of the `TransformerLayer` is going to be anyway in most cases. You just have to do the calculation in your head, which seems wasteful :)

**Comment by skaae** [comment]
The motivation was that i wanted to test if it worked to downsample the output from the transformerlayer by e.g factor 2 and 4.

I agree that scale is a better name. The output shape is calculated from the input shape, but height and width has to be fixed otherwise the location network will not work i think?

Are you playing around with the layer?

**Comment by benanne** [comment]
I see. But currently the input is not being rescaled before the transform is applied, right? I just think it's a bit strange outside of that context, my thought process would be "okay, I have 60x60 input images, and I want the output of the TransformerLayer to be 20x20, so I should use a scale factor of 3". In that sense it's much more natural to just be able to specify (20, 20), because that is what I want.

Things get hairier when your input is 79x54 and you want a 23x23 output, for example :)

**Comment by skaae** [comment]
No. The downsample_factor just determines how many "pixels" to sample.
It's fine by me to change downsample factor to height and width. 

I can't come up with a common usecase where the input size would change during training.

**Comment by f0k** [comment]
But what if you stack two of those things on top of each other (with some convolutions in between) and then end in a dense layer? It's wasteful to keep track of the transformer input shapes in your head than, it's easier to think in terms of factors. So I'd prefer to keep both!

**Comment by benanne** [comment]
> But what if you stack two of those things on top of each other (with some convolutions in between) and then end in a dense layer? It's wasteful to keep track of the transformer input shapes in your head than, it's easier to think in terms of factors. So I'd prefer to keep both!

I don't follow, which two things? Even if you end in a dense layer, you still need a shape that remains constant across examples. The only situation where it would make sense is if you stacked convolutions/pooling and then had a global pooling layer at the end. But I have a feeling there's a misunderstanding here.

**Comment by f0k** [comment]
> I agree that scale is a better name.

Note that this would also mean different semantics. At least for me `scale=2` would imply upscaling the image, not downscaling.

> The output shape is calculated from the input shape, but height and width has to be fixed otherwise the location network will not work i think?

The localization network doesn't have to use the same input as the transformer network, e.g., it could use a version scaled to a fixed size, or different features altogether.

> Even if you end in a dense layer, you still need a shape that remains constant across examples. [...] But I have a feeling there's a misunderstanding here.

Yes, we're mixing two things here :)
a) Do we need to support a scaling factor at all, or is an output size enough?
b) Is there any benefit from computing the output size symbolically, from a factor given at construction time?

I was referring to your following comment, which I interpreted to address the first point:

> So I don't really understand how being able to specify a scale factor exempts you from knowing the input shape -- you still need to know what shape the output of the TransformerLayer is going to be anyway in most cases.

If the TransformerLayer is not the first thing in your pipeline, then you don't necessarily know the input shape, and you don't necessarily have a particular output shape in mind, but maybe you have an idea about what would be a good factor (or you can try a few). So I'd like to keep support for a scale factor.

About the second point, you wrote:

> But a fixed shape is what you want, no? Except in the special case where the network you stack on top of it is fully convolutional and you have some kind of global pooling layer at the end.

Yes... I agree that's a very unlikely use case. The TransformerLayer might actually be more useful to bring images from a variable input size to a fixed output size (although that would limit the ability to do batch processing). But apart from a lack of use cases, is there a good reason for not supporting the combination of `input_shape == (batch_size, channels, None, None)` and a factor? It doesn't really complicate the code if we agree we want to support both factor and size specifications.

**Comment by benanne** [comment]
> If the TransformerLayer is not the first thing in your pipeline, then you don't necessarily know the input shape, and you don't necessarily have a particular output shape in mind, but maybe you have an idea about what would be a good factor (or you can try a few). So I'd like to keep support for a scale factor.

I don't get this, in fact I would think it's the other way around: maybe you don't know the exact input shape, but you know roughly what you want the output shape to be. You need to know what the output shape of the TransformerLayer is going to be anyway, to be able to know what you can stack on top of it.

Anyway, I don't object to keeping support for a scale factor, there must be enough other people who prefer it that way :)

**Comment by f0k** [comment]
> You need to know what the output shape of the TransformerLayer is going to be anyway, to be able to know what you can stack on top of it.

That's just because you always take care not to pool things that don't evenly divide :) Even then you can always stack some 3x3 convolutions or a dense layer on top, can't you?

> Anyway, I don't object to keeping support for a scale factor, there must be enough other people who prefer it that way :)

Good! Point a) settled then! What about point b)?

**Comment by benanne** [comment]
regarding b), I guess it's better to support it since it doesn't really add any overhead, does it? Nor does it make the code any more difficult to understand (or does it).

I can't really think of any practical use cases... you could have a model that supports variable input sizes (as I said, for example by having a global average pooling layer at the top), but if you add a `TransformerLayer` in that you would probably to do it to specifically focus on a fixed-size region in the image, I guess. So I don't know how realistic that is.

**Comment by f0k** [comment]
> regarding b), I guess it's better to support it since it doesn't really add any overhead, does it? Nor does it make the code any more difficult to understand (or does it).

Not a lot. It's just that if we don't support it, we can handle the distinction between `scale` and `size` it in the constructor and set `self.size` to an integer tuple in either case. If we support an unknown input size paired with a factor, we'll have to defer the distinction to `get_output_shape_for()` and `get_output_for()`. It's a minor annoyance...

**Comment by benanne** [comment]
I see. In that case maybe we shouldn't support it. Maybe let's see if anyone can come up with a plausible use case within the next two days or so :p
